RUFAS/biophysical/animal/herd_manager.py

```diff
@@ -1,2 +1,2079 @@
+        [35, 21, 14]
+
+        >>> HerdManager._plan_animal_allocation(num_animals=47, max_spaces_in_pens=[50, 30, 20])
+        [22, 15, 10]
+
+        """
+        num_pens_for_combination = len(max_spaces_in_pens)
+        overall_density = self._calculate_density(num_animals=num_animals, num_spaces=sum(max_spaces_in_pens))
+
+        if overall_density > 1.0:
+            raise ValueError("The number of animals cannot exceed the total number of spaces.")
+
+        num_animals_in_pens = [0] * num_pens_for_combination
+        allocation_limits = [math.ceil(overall_density * max_spaces) for max_spaces in max_spaces_in_pens]
+
+        sorted_pen_indices = sorted(
+            range(num_pens_for_combination),
+            key=lambda pen_idx: (allocation_limits[pen_idx], pen_idx),
+        )
+
+        for i in sorted_pen_indices[: num_pens_for_combination - 1]:
+            num_animals_to_allocate = min(num_animals, allocation_limits[i])
+            num_animals_in_pens[i] = num_animals_to_allocate
+            num_animals -= num_animals_to_allocate
+        num_animals_in_pens[sorted_pen_indices[-1]] += num_animals
+
+        return num_animals_in_pens
+
+    def _execute_allocation_plan(
+        self,
+        allocation_plan: list[int],
+        animals: list[Animal],
+        animal_pens: list[Pen],
+    ) -> None:
+        """
+        Execute an allocation plan to distribute animals into pens according to the given plan.
+
+        This method iterates over the provided allocation plan and updates each pen with the specified number
+        of animals.
+
+        Parameters
+        ----------
+        allocation_plan : list[int]
+            A list of integers representing the number of animals to be allocated to each pen.
+            The length of the allocation_plan list must match the number of pens in animal_pens.
+        animals : list[Animal]
+            A list of animals to be allocated among the pens.
+        animal_pens : list[Pen]
+            A list of Pen objects representing the pens to which animals will be allocated.
+
+
+        Returns
+        -------
+        None
+
+        Raises
+        ------
+        ValueError
+            If the length of the allocation plan does not match the number of pens.
+            If the sum of the allocation plan does not match the number of animals.
+
+        """
+
+        if len(allocation_plan) != len(animal_pens):
+            raise ValueError("The length of the allocation plan must match the number of pens.")
+        elif sum(allocation_plan) != len(animals):
+            raise ValueError("The sum of the allocation plan must match the number of animals.")
+
+        for i, count in enumerate(allocation_plan):
+            animal_pens[i].insert_animals_into_animals_in_pen_map(animals[:count])
+            animals = animals[count:]
+
+    def _sort_cows_before_allocation(self) -> None:
+        """
+        Sort lactating cows by days_in_milk in increasing order.
```

Comment:
```suggestion
        Sort cows by days_in_milk in increasing order.
```
I think this is a little more correct because this includes dry cows